### ISSUE TITLE
[Bug]: 

### ISSUE THREAD

**Issue by marcel-mattr** [hashicorp/terraform-provider-aws#29323]
### Terraform Core Version

1.3.7

### AWS Provider Version

4.53

### Affected Resource(s)

* aws_flow_log

### Expected Behavior

Following the given [example](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/flow_log#cloudwatch-logging) on AWS Flow Log documentation, when I try to create this resource using Terraform, I got an error described below.

### Actual Behavior

```
module.vpc.aws_flow_log.this[0]: Creating...
╷
│ Error: creating Flow Log (vpc-XXXXXXXXX): InvalidParameter: Unable to assume given IAM role: 'arn:aws:iam::ACCOUNT_ID:role/vpc-flow-log-role-9999999999999'
│       status code: 400, request id: xxxxxxxx-yyyy-xxxx-yyyy-xxxxxxxxxxxxx
│
│   with module.vpc.aws_flow_log.this[0],
│   on ../modules/aws/vpc/aws_vpc_flow_log.tf line 110, in resource "aws_flow_log" "this":
│  110: resource "aws_flow_log" "this" {
│
╵
```

### Terraform Configuration Files

```
resource "aws_flow_log" "example" {
  iam_role_arn    = aws_iam_role.example.arn
  log_destination = aws_cloudwatch_log_group.example.arn
  traffic_type    = "ALL"
  vpc_id          = aws_vpc.example.id
}

resource "aws_cloudwatch_log_group" "example" {
  name = "example"
}

resource "aws_iam_role" "example" {
  name = "example"

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "Service": "vpc-flow-logs.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF
}

resource "aws_iam_role_policy" "example" {
  name = "example"
  role = aws_iam_role.example.id

  policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": [
        "logs:CreateLogGroup",
        "logs:CreateLogStream",
        "logs:PutLogEvents",
        "logs:DescribeLogGroups",
        "logs:DescribeLogStreams"
      ],
      "Effect": "Allow",
      "Resource": "*"
    }
  ]
}
EOF
}
```

### Steps to Reproduce

Set up a VPC Flow Log on a VPC using the [documentation example](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/flow_log#cloudwatch-logging).

### Would you like to implement a fix?

No

**Comment by djlambert** [comment]
This might be related https://stackoverflow.com/a/53183325.

I created my flow log through the console and imported it. `terraform plan` shows: 
```
  # module.network.aws_flow_log.vpc must be replaced
-/+ resource "aws_flow_log" "vpc" {
      ~ arn                      = "arn:aws:ec2:us-east-2:123456789:vpc-flow-log/fl-0751b9c0b4d94da72" -> (known after apply)
      ~ id                       = "fl-0751b9c0b4d94da72" -> (known after apply)
      + log_destination          = "arn:aws:logs:us-east-2:123456789:log-group:/aws/vpc/dev-primary" # forces replacement
      ~ log_format               = "${version} ${account-id} ${interface-id} ${srcaddr} ${dstaddr} ${srcport} ${dstport} ${protocol} ${packets} ${bytes} ${start} ${end} ${action} ${log-status}" -> (known after apply)
      ~ log_group_name           = "/aws/vpc/dev-primary" -> (known after apply)
      - tags                     = {} -> null
      ~ tags_all                 = {} -> (known after apply)
        # (5 unchanged attributes hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```
If I instead specify the deprecated `log_group_name` argument instead of `log_destination`, `terraform plan` gives me a deprecation warning but says the configuration matches.

**Comment by marcel-mattr** [comment]
@djlambert I just found out a weird thing.

I tried a workaround on this bug using `null provider` and `aws cli`. If I add a sleep timer between creating the IAM rule and it worked.

```tf
resource "null_resource" "iam_role" {
  triggers = {
    profile     = var.profile
    role_name   = "${var.vpc_flow_log_role_name}"
    policy_name = "${var.vpc_flow_log_policy_name}"
  }

  provisioner "local-exec" {
    when    = create
    command = <<-EOT
      aws iam create-role --profile ${self.triggers.profile} --role-name ${self.triggers.role_name} --assume-role-policy-document file://${path.module}/data/iam-assume-policy.json
      aws iam put-role-policy --profile ${self.triggers.profile} --role-name ${self.triggers.role_name} --policy-name ${self.triggers.policy_name} --policy-document file://${path.module}/data/iam-policy.json
      sleep 30
    EOT
  }

  provisioner "local-exec" {
    when    = destroy
    command = <<-EOT
      aws iam delete-role-policy --profile ${self.triggers.profile} --role-name ${self.triggers.role_name} --policy-name ${self.triggers.policy_name}
      aws iam delete-role --profile ${self.triggers.profile} --role-name ${self.triggers.role_name}
    EOT
  }
}

resource "aws_cloudwatch_log_group" "vpc_flow_logs" {
  name  = "${var.vpc_flow_log_group_name}"
}

resource "aws_flow_log" "vpc_flow_logs" {
  iam_role_arn         = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/${var.vpc_flow_log_role_name}"
  log_destination      = aws_cloudwatch_log_group.vpc_flow_logs.arn
  log_destination_type = "cloud-watch-logs"
  traffic_type         = coalesce(var.vpc_flow_log_traffic_type, "ALL")
  vpc_id               = local.flowlog_vpc_id
}
```

As you can see on your own, if I use `sleep N` above 15 seconds, it works.

**Comment by Ankita5892** [comment]
Same issue I am also facing, giving delay is not a good solution for production script, anyone one why suddenly it’s start giving error same script was working fine till 8 February 2023

**Comment by sblask** [comment]
Got the same problem and a retry fixes it. It's like the assume role policy isn't immediately propagated. But as it's part of the create role call I can't add a lifecycle policy. Could be an AWS bug?

**Comment by marcel-mattr** [comment]
@Ankita5892 I totally agree that adding a delay is not a good solution but till they solve this issue on their side, it is what it is.

@sblask it is because I tried to create the same resource using terraform and aws-cli. Without the delay, it breaks.

**Comment by jmayhill1** [comment]
My team uses slightly earlier versions of Terraform Core and the AWS Provider and see the same results as described in the original post. We saw a few sporadic failures in the last few months, but now it is almost every request.  We're looking at the local exec sleep "hack", but thinking about trying depends_on first as in the example here: https://developer.hashicorp.com/terraform/language/v1.1.x/meta-arguments/depends_on 
Has anyone tried it?

**Comment by iusetabs** [comment]
Also facing this issue, however I am using an application which generates terraform configurations and there doesn't seem to be a way using this application to add a sleep command. So if anybody else has a suggestion for how to work around this, I would appreciate it! 

Also +1 to the suggestion that there seems to be a bug on the AWS side. Anybody know where we can open a bug to get this checked?

**Comment by Ankita5892** [comment]
@jmayhill1 i tried with depends on in flow logs resource for iam role and policy but it’s still failed with same error so depends on inside the resource not working. 
I have raised ticket with aws support but they are also not able to answer as from console we are able to create but still waiting for feedback from aws, if anyone know any alternate way please suggest.

**Comment by jmayhill1** [comment]
@Ankita5892 Thanks for the confirmation.  We've also tried variations of the depends on with it not working for us either. I suspect that something on the AWS side has changed, because nothing in our code has changed, but it could have been terraform relying on AWS timing which is not reliable. We saw a gradual increase in the number of failures with this error until now we see the error consistently.  Although the example code shows separate resources for the role and the inline policy is there anyway to define the policy inline like the assume role policy?

**Comment by marcel-mattr** [comment]
@jmayhill1 it's the code used in the documentation

**Comment by present-lee** [comment]
I tried with depends_on option in VPC module.

1. create iam role and policy, cloudwatch log group
2. create vpc with depends_on option
    resource "aws_vpc" "main" {
        cidr_block = var.vpc_cidr

        enable_dns_hostnames = true

        depends_on = [
            aws_cloudwatch_log_group.flow_log_group,
            aws_iam_role.flow_log_role,
           aws_iam_role_policy.flow_log_policy
       ]
    }
3. create aws_flow_log with deponds_on option
    resource "aws_flow_log" "vpc_flow_log" {
      iam_role_arn    = aws_iam_role.flow_log_role.arn
      log_destination = aws_cloudwatch_log_group.flow_log_group.arn
      traffic_type    = "ALL"
      vpc_id          = aws_vpc.main.id

       depends_on = [
            aws_iam_role.flow_log_role,
            aws_iam_role_policy.flow_log_policy,
            aws_vpc.main
        ]
    }

it works good to me.
The reason I think is that while the vpc is being created, the iam role can be fully created.

**Comment by Ankita5892** [comment]
Hi @iusetabs @jmayhill1 , there is api delay from aws side due to some security POV ( which recently they did) so if you are using module, you can keep one separate module for flow Logs IAM role and policy and one for flow logs that’s how I am creation will have sufficient time to create and arn you can attach to flow logs
